tket/src/WeightSubgrMono/include/WeightSubgrMono/Searching/SearchBranch.hpp

```diff
@@ -0,0 +1,147 @@
+// Copyright 2019-2022 Cambridge Quantum Computing
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+#include <map>
+#include <optional>
+#include <set>
+#include <string>
+
+#include "../WeightPruning/WeightNogoodDetectorManager.hpp"
+#include "EnrichedNode.hpp"
+
+namespace tket {
+namespace WeightedSubgraphMonomorphism {
+
+struct FixedData;
+struct SharedData;
+
+/** Represents a depth-first traversal.
+ * However, it's really about reducing nodes etc.,
+ * not the search logic. This contains the list of nodes so far,
+ * which may be viewed as a stack, but the caller is responsible for
+ * moving up and down the list.
+ */
+class SearchBranch {
+ public:
+  SearchBranch();
+
+  /** Fills in the top node, but might not FULLY reduce.
```

Comment:
A search branch (or search tree) consists of a vector of nodes. Each node contains all Dom(PV), i.e. the possible TV to which PV can be assigned. To search, we choose a PV for which Dom(PV) has size > 1, choose x in Dom(PV), then move down one level to a new node and assign PV->x, then calculate further ("reduce the node").

However, obviously the first node (AKA top node, or node[0]) has to be filled in to start.

There is some complication because I was trying multiple parallel search branches, so we need to store the INITIAL top node before any searching takes place, and then copy that to NEW branches (we can't just copy the top node from other branches because it changes as the search progresses and backtracks); but since experiments showed parallel searching to be not useful, I'm just going to clear all that out and simplify, with just ONE branch.

Reply:
Thanks for the explanation - that makes sense.